### ISSUE TITLE
PouchDB LiveFind discussion

### ISSUE THREAD

**Issue by colinskow** [nolanlawson/pouchdb-find#181]
As a front-end app developer, I want the UI to be in sync with database changes and I want that to happen with the least possible coding on my part. The main problem with CouchDB queries is they don't update themselves when the database changes.

I want to be able to subscribe to a query and see updates every time a change impacts the results.

It is fairly simple to do a `db.allDocs()`, subscribe to the changes feed, and keep an up-to-date array of all the docs in the database. But running a given query there is no way of knowing if a change matches that query, without re-implementing the query system from scratch.

So in order to implement this I need a way to know if a doc in the changes feed matches the original query.

I want to build a PouchDB plugin that depends on `pouchdb-find` to implement this. Here is the usage I have envisioned:

```js
db.liveFind({
  selector: {
    series: 'Mario',
    debut: { $gt: 1990 }
  },
  sort: ['debut'],
  throttle: 250
}).on('update', (latestResults) => {
  updateUI(latestResults.docs);
});
```

What I need is a way to check changes against the selector as they come in. I also need the sort algorithm to keep the list in the right order. I would like to see if there is a simple way to access the internal code of `pouchdb-find` to accomplish this rather than re-invent the wheel.

I've looked at a library called [Sift.js](https://github.com/crcn/sift.js) which implements MongoDB style selectors on an in-memory array of docs, but there may be subtle differences in implementation. (For example sift.js lacks the CouchDB collation engine.)

@nolanlawson @garrensmith any advice would be appreciated.

**Comment by garrensmith** [comment]
@colinskow I think something like this has just landed in CouchDB https://issues.apache.org/jira/browse/COUCHDB-2988 the next step would be to integrate it into PouchDB.

**Comment by colinskow** [comment]
@garrensmith That's great news! The only possible problem I see with that implementation is how to know when a doc that used to match the query changes and no longer matches. CouchDB would need to remember the previous changes it has sent me and alert me when I need to remove a doc from the list.

Right now my idea is to watch all changes and test each one against the selector. This is simple on a client implementation like PouchDB, but not completely efficient receiving traffic from a remote server.

**Comment by colinskow** [comment]
Good news... I got it working!!! And it was only a couple dozen lines of code. I plugged directly into `adapter/local/find/in-memory-filter.js` and I filter each change through the in-memory query there.

I'm working on writing tests now to make sure everything performs as expected. This is going to make it a lot easier to write front-end apps with PouchDB and the UI will effortlessly stay in sync as the DB changes.
